src/phongo_bson.c

```diff
@@ -1037,6 +1087,62 @@ bool php_phongo_bson_to_zval_ex(const unsigned char* data, int data_len, php_pho
 		case PHONGO_TYPEMAP_CLASS: {
 			zval obj;
 
+#if PHP_VERSION_ID >= 80100
+			/* Note: this likely only works with backed enums */
+			if (state->odm && state->odm->ce_flags & ZEND_ACC_ENUM) {
+				zval *value = php_array_fetchc(&state->zchild, "value");
+				zval *retval;
+
+				if (!value) {
+					char* path = php_phongo_field_path_as_string(state->field_path);
+					phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE, "Missing enum value for field path '%s'", path);
+					efree(path);
+
+					goto cleanup;
+				}
+
+				retval = zend_call_method_with_1_params(NULL, state->odm, NULL, "from", NULL, value);
```

Comment:
@sgolemon: while the approach below seems like it'd be portable for both pure and backed enums, I wanted to see if invoking [`BackedEnum::from()`](https://www.php.net/manual/en/language.enumerations.backed.php) would work to reduce internal code.

Running this code path in the corresponding test, I see the following error:

```
Fatal error: MyEnum::from(): Return value must be of type static, MyEnum returned in Unknown on line 0
```

I traced this error back to either `zend_verify_return_error` or `zend_verify_internal_return_error` in [`zend_execute.c`](https://github.com/php/php-src/blob/php-8.1.5/Zend/zend_execute.c#L1306), but it's not clear to me why this would trigger.

I'm passing the enum's CE in `zend_call_method_with_1_params`. It wasn't possible to pass a `zend_object*` as the first parameter since enums cannot be instantiated.

I find it odd that I wouldn't be able to invoke these methods from an internal context (presumably I should be able to do everything one can do in userland), so I wonder if I may just be hitting an edge case that may be a bug in the original enum implementation. Any ideas?

Reply:
The good news is: What you're trying to do SHOULD work.  The bad news is that there seems to be a bug in the engine where we make an assumption during return type checking that we have a relevant call frame to look at when verifying the called scope.  Turns out for Enums, which have an internal implementation of the from/tryFrom methods, we don't actually have a call frame, so the scope is essentially meaningless.

I've got a hack you can use, which violates the call stack by replacing whatever it's current notion of "thisness" with the actual calling scope you want enforced, and the rest "just works".  I wouldn't recommend this in a general use case, but for calling Enum::from() it should be safe enough.

```
ZEND_FUNCTION(smg)
{
    zend_string *classname = zend_string_init("ZendTestStringEnum", strlen("ZendTestStringEnum"), 0);
    zend_class_entry *ce = zend_lookup_class(classname);
    ZEND_ASSERT(ce);

    zval value;
    ZVAL_STRING(&value, "Test1");

    /* Preserve current scope and replace with target scope. */
    zval oldThis = EG(current_execute_data)->This;
    ZVAL_CE(&(EG(current_execute_data)->This), ce); 

    zend_call_method_with_1_params(NULL, ce, NULL, "from", return_value, &value);

    /* Restore current scope */
    EG(current_execute_data)->This = oldThis;

    zend_string_release(classname);
    zval_ptr_dtor(&value);
}
```

The two commented sections being what you wanna pay attention to.